dev/NavigationView/NavigationView.cpp

```diff
@@ -1594,6 +1734,222 @@ void NavigationView::UpdateVisualStateForDisplayModeGroup(const winrt::Navigatio
+            args.Handled(true);
+            KeyboardFocusFirstItemFromItem(nvi);
+            break;
+        case winrt::VirtualKey::End:
+            args.Handled(true);
+            KeyboardFocusLastItemFromItem(nvi);
+            break;
+        }
+    }
+}
+
+void NavigationView::KeyboardFocusFirstItemFromItem(const winrt::NavigationViewItemBase& nvib)
+{
+    auto const firstElement = [this, nvib]()
+    {
+        if (IsTopNavigationView())
+        {
+            if (IsContainerInOverflow(nvib))
+            {
+                return m_topNavRepeaterOverflowView.get().TryGetElement(0);
+            }
+            else
+            {
+                return m_topNavRepeater.get().TryGetElement(0);
+            }
+        }
+        return m_leftNavRepeater.get().TryGetElement(0);
+    }();
+
+    if (auto controlFirst = firstElement.try_as<winrt::Control>())
+    {
+        controlFirst.Focus(winrt::FocusState::Keyboard);
+    }
+}
+
+void NavigationView::KeyboardFocusLastItemFromItem(const winrt::NavigationViewItemBase& nvib)
+{
+    auto const ir = [this, nvib]()
+    {
+        if (IsTopNavigationView())
+        {
+            if (IsContainerInOverflow(nvib))
+            {
+                return m_topNavRepeaterOverflowView.get();
+            }
+            else
+            {
+                return m_topNavRepeater.get();
+            }
+        }
+        else
+        {
+            return m_leftNavRepeater.get();
+        }
+    }();
+
+    if (auto itemsSourceView = ir.ItemsSourceView())
+    {
+        auto lastIndex = itemsSourceView.Count() - 1;
+        if (auto lastElement = ir.TryGetElement(lastIndex))
+        {
+            if (auto controlLast = lastElement.try_as<winrt::Control>())
+            {
+                controlLast.Focus(winrt::FocusState::Keyboard);
+            }
+        }
+    }
+}
+
+void NavigationView::RepeaterGettingFocus(const winrt::IInspectable& sender, const winrt::GettingFocusEventArgs& args)
+{
+    auto const inputDevice = args.InputDevice();
+    if (inputDevice == winrt::FocusInputDeviceKind::Keyboard)
+    {
+        if (auto const oldFocusedElement = args.OldFocusedElement())
+        {
+            auto const oldElementParent = winrt::VisualTreeHelper::GetParent(oldFocusedElement);
+            auto const rootRepeater = [this]()
+            {
+                if (IsTopNavigationView())
+                {
+                    return m_topNavRepeater.get();
+                }
+                return m_leftNavRepeater.get();
+            }();
+            // If focus is coming from outside the root repeater, put focus on last focused item
+            if (rootRepeater != oldElementParent)
+            {
+                if (auto const lastFocusedNvi = rootRepeater.TryGetElement(m_indexOflastFocusedItem))
+                {
+                    if (auto const argsAsIGettingFocusEventArgs2 = args.try_as<winrt::IGettingFocusEventArgs2>())
+                    {
```

Comment:
I think swaping the order of these if statements would be better, I think tryGetElement will succeed more often than the version checking cast.